Adding example of cache management methods usage

langsmith opened the pull request:
**Don't merge this pr to `master` until the stable 8.2.0 release of the Maps SDK for Android goes out!**

Ok @mapbox/maps-android (cc @jmkiley too), this pr resolves https://github.com/mapbox/mapbox-android-demo/issues/1015 by hooking [the Maps SDK's new cache management methods](https://github.com/mapbox/mapbox-gl-native/pull/14962) into 3 different options:

- Server-side message from Firebase Cloud Messaging
- Firebase Remote Config
- Android XML UI

It was pretty brutal to get the Firebase libraries, Gradle, Firebase cloud testing, CircleCI, and the app's flavors, to all play nicely with one other. 

*The main addition of this pr is `CacheManagementActivity`.* Its XML UI features a button to trigger each of the new methods. This is similar to what's done [in @tobrun's Maps SDK test app example](https://github.com/mapbox/mapbox-gl-native/pull/14962/files#diff-41f5fa1367092e4c80d10e526a13eb04). 

![ezgif com-resize (1)](https://user-images.githubusercontent.com/4394910/61349207-22eab780-a818-11e9-9751-6171729d5245.gif)


The activity also shows how to use [Firebase's Remote Config tool](https://firebase.google.com/docs/remote-config/use-config-android). 

<img width="430" alt="Screen Shot 2019-07-15 at 11 54 35 AM" src="https://user-images.githubusercontent.com/4394910/61257673-e5f2c800-a726-11e9-9785-84732230a7d0.png">

Once key/value pairs are added to the Firebase console, the console can be used to update the value. Whenever the activity starts, the key/values are checked on the device. In `CacheManagementActivity`, they're checked before the `MapView` is set up. In the screenshot above of this app's remote config console, the ambient cache would be cleared before the `MapView` initializes.

*`CustomFirebaseMessagingService` is the other major part of this pr.* It receives inbound info from a backend server. 

The developer/company will [need to set up a server-side app via the Firebase Admin SDK or another method](https://firebase.google.com/docs/cloud-messaging/server#firebase-admin-sdk-for-fcm). 

The Firebase Admin SDK isn't meant for an Android app, which is why this pr doesn't wade into Firebase servers, admin, accounts, etc.

  - https://stackoverflow.com/questions/42097740/firebase-admin-sdk-for-android-methods-not-found#comment71366069_42097740
  - https://stackoverflow.com/a/42103063/6358488

Eventually, the server will send [a Firebase `Data message`](https://firebase.google.com/docs/cloud-messaging/concept-options#notifications_and_data_messages) with key/value pairs, which will represent whether the cache should be adjusted in a certain way.

<img width="601" alt="Screen Shot 2019-07-15 at 5 34 54 PM" src="https://user-images.githubusercontent.com/4394910/61257695-facf5b80-a726-11e9-8517-ccd826b65915.png">

The value of the key/value pair can be parsed in `CustomFirebaseMessagingService`. For example:

```java
boolean clearAmbientCache = parseBoolean(data.get("clearAmbientCache"));
```

The resulting boolean can then determine whether the corresponding Maps SDK cache method should be called.

Last but not least, I create the `fabric.properties` file at https://github.com/mapbox/mapbox-android-demo/compare/ls-firebase-cache-setup?expand=1#diff-29944324a3cbf9f4bd0162dfe3975d88R60 because copy/pasting the local machine version of the raw `fabric.properties` file as a CircleCI environmental variable couldn't figure out how to include the break. I do `echo "{GOOGLE_SERVICES_JSON_RELEASE}" > MapboxAndroidDemo/google-services.json` to make the JSON file, but the equivalent didn't work for making the `fabric.properties` file in one go. 

When I try to turn the environment variable into the `fabric.properties` file, I would get the following error:

<img width="945" alt="Screen Shot 2019-07-16 at 4" src="https://user-images.githubusercontent.com/4394910/61347376-96d59180-a811-11e9-92d8-83fdb12013e9.png">

I finally realized that the error probably stemmed from a missing line break. There is https://circleci.com/docs/2.0/env-vars/#encoding-multi-line-environment-variables but I wasn't sure how to get it to create a two-line file in the following required format:

```
apiSecret=XXXX
apiKey=XXXX
```

Would my current `fabric.properties` setup block others/non-Mapboxers from getting pull request CI tests to pass, as mentioned by @tobrun at https://github.com/mapbox/mapbox-android-demo/pull/1104#pullrequestreview-258706626 ? If so, I'm not sure how. CircleCI would block their test from accessing the saved environment variables?

langsmith commented:
Here's a caveat about remote config usage, which is worth mentioning for visibility. 

As written in https://firebase.google.com/docs/remote-config/android#throttling, the default minimum fetch interval for remote config is 12 hours. No matter how many fetch calls are actually made, the firebase server-side config setup won't be fetched more than once in a 12-hour window. 

![Screen Shot 2019-07-17 at 1 03 31 PM](https://user-images.githubusercontent.com/4394910/61407748-4dc62180-a893-11e9-826f-1244337bd975.png)

This throttling could affect how quickly the cache adjustment changes roll out to devices if remote config is used.

I've added a code comment on this pr and commented out the code that sets a custom fetch interval.

zugaldia commented:
> I finally realized that the error probably stemmed from a missing line break. There is https://circleci.com/docs/2.0/env-vars/#encoding-multi-line-environment-variables but I wasn't sure how to get it to create a two-line file in the following required format:

Didn't the base64 encoding approach work in this case?

> As written in https://firebase.google.com/docs/remote-config/android#throttling, the default minimum fetch interval for remote config is 12 hours. No matter how many fetch calls are actually made, the firebase server-side config setup won't be fetched more than once in a 12-hour window.

This means that remote config wouldn't be necessarily the best option to issue a remote cache invalidation because the server default caching time for Mapbox Maps API is of the same order of magnitude. However, it still makes it a valid mechanism to update other parameters, like the total cache size.

Finally, removing the `DO NOT MERGE` tag now that 8.2.0 has landed.

LukasPaczos commented on MapboxAndroidDemo/src/global/java/com/mapbox/mapboxandroiddemo/MapboxApplication.java:
```diff
@@ -40,10 +41,6 @@ private void setUpTileLoadingMeasurement() {
-            .disabled(BuildConfig.DEBUG)
-            .build())
-        .build());
+    Fabric.with(this, new Crashlytics());
```
Does this mean that we'll send reports from the debug versions as well? I think we should stick to release builds only.

langsmith replied:
Unfortunately, setting this to debug starts a cascade of other build issues because of all of the various Firebase libraries that are now in the app.

The Crashlytics section of the Firebase console is set up to release vs. debug anyway https://console.firebase.google.com/project/mapbox-android-demo/crashlytics. I've set up CircleCI with the release version of the google-services.json file, so that release crashes are sent to the right dashboard.

langsmith replied:
Looks like I found a workaround

https://github.com/mapbox/mapbox-android-demo/pull/1139#issuecomment-517469922

LukasPaczos commented on MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/offline/CacheManagementActivity.java:
```diff
@@ -0,0 +1,284 @@
+  }
+
+  /**
+   * Invalidate the ambient cache. See ANDROID-DOCS-URL-HERE for more
```
Missing the doc? :)
Same for below occurrences.

langsmith replied:
Fixed. All changed to https://docs.mapbox.com/android/maps/overview/offline/#cache-management, which will be the correct URL when https://github.com/mapbox/android-docs/pull/995 lands.

LukasPaczos commented on MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/offline/CacheManagementActivity.java:
```diff
@@ -0,0 +1,284 @@
+            findViewById(R.id.set_max_ambient_cache_size_button).setOnClickListener(new View.OnClickListener() {
+              @Override
+              public void onClick(View view) {
+                EditText newMaxAmbientCacheSizeTextView = findViewById(
```
NIT: `findViewById` calls are generally expensive and shouldn't be used in recurring scenarios (button click). They should rather be cached.

langsmith replied:
Fixed

LukasPaczos commented on MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/offline/CacheManagementActivity.java:
```diff
@@ -0,0 +1,284 @@
+
+      @Override
+      public void onError(@NonNull String message) {
+        showToast(R.string.cache_invalidated_toast_error);
```
Would be nice to pass the error message tom the toast as well.
Same for below occurrences.

langsmith replied:
Done

LukasPaczos commented on MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/offline/CacheManagementActivity.java:
```diff
@@ -0,0 +1,284 @@
+
+  private void showToast(int toastStringId) {
+    // validate that all callbacks occur on main thread
+    if (Looper.myLooper() == Looper.getMainLooper()) {
```
This will not show a toast if this method is invoked from a worker thread. Could you try additionally post to the main thread?

langsmith replied:
would the following work?

```java
if (Looper.myLooper() == Looper.getMainLooper()) {
      Toast.makeText(CacheManagementActivity.this,
          textToShow, Toast.LENGTH_SHORT).show();
    } else {
      CacheManagementActivity.this.runOnUiThread(new Runnable() {
        public void run() {
          Toast.makeText(CacheManagementActivity.this,
              textToShow, Toast.LENGTH_SHORT).show();
        }
      });
}
```

LukasPaczos replied:
`Activity#runOnUiThread` runs the code immediately if you're on UiThread already (basically performs the same check you did), so just
```
      CacheManagementActivity.this.runOnUiThread(new Runnable() {
        public void run() {
          Toast.makeText(CacheManagementActivity.this,
              textToShow, Toast.LENGTH_SHORT).show();
        }
      });
```
is enough.

langsmith replied:
Ok, fixed.

LukasPaczos commented on MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/utils/CustomFirebaseMessagingService.java:
```diff
@@ -0,0 +1,82 @@
+ * With the right server side notification structure, the Maps SDK cache management methods
+ * can be hooked into the inbound notification data.
+ */
+public class CustomFirebaseMessagingService extends FirebaseMessagingService {
```
Is this class practically used here?

langsmith replied:
No, it's not. Will add a note that it's simply here as a solid starting point to finishing a serverside notification setup.

langsmith commented:
Ok @LukasPaczos , https://github.com/mapbox/mapbox-android-demo/pull/1139/commits/cb11790fb638f0a4488d3a1244592be5dd560a04 addresses your comments above. Would love another round of review when you get the chance. Thanks.

LukasPaczos commented:
I'm running into `File google-services.json is missing. The Google Services Plugin cannot function without it.` when trying to run the project. Can we include a debug json or does it contain private info?

langsmith commented:
> Can we include a debug json or does it contain private info?

Contains private info that can't be pushed to a public repo. I'm looking into remedies.

langsmith commented:
> I'm running into File google-services.json is missing. The Google Services Plugin cannot function without it. when trying to run the project. Can we include a debug json or does it contain private info?

@LukasPaczos ,  https://github.com/mapbox/mapbox-android-demo/pull/1139/commits/5d5787042103cee86ab278ab455d6200622f4758 makes some tweaks to get around the `google-services.json` requirement. It also disables Crashlytics for debug. With this latest commit, the app installs fine on my device and I can successfully open/use the `CacheManagementActivity`. Hopefully the same is for you on your device. Please check it out when you get a chance.

LukasPaczos commented on MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/offline/CacheManagementActivity.java:
```diff
@@ -0,0 +1,299 @@
+            findViewById(R.id.set_max_ambient_cache_size_button).setOnClickListener(new View.OnClickListener() {
+              @Override
+              public void onClick(View view) {
+                long newMaxAmbientCacheSize = Long.valueOf(
```
If the edit text is empty, this will throw a `NumberFormatException`.

langsmith replied:
Ok, fixed to

```java
 findViewById(R.id.set_max_ambient_cache_size_button).setOnClickListener(new View.OnClickListener() {
              @Override
              public void onClick(View view) {
                try {
                  long newMaxAmbientCacheSize = Long.valueOf(
                      newMaxAmbientCacheSizeTextView.getText().toString());
                  setNewMaxAmbientCache(newMaxAmbientCacheSize);
                } catch (NumberFormatException exception) {
                  Logger.d("CacheManagementActivity", "%s", exception);
                }
              }
            });
```

LukasPaczos commented on MapboxAndroidDemo/src/main/res/layout/activity_offline_cache_manipulation.xml:
```diff
@@ -0,0 +1,98 @@
+        android:id="@+id/set_max_ambient_cache_size_button_editText"
+        android:layout_width="100dp"
+        android:layout_height="wrap_content"
+        android:hint="@string/set_ambient_cache_max_hint"
```
NIT: this can specify `android:inputType="number"`.

langsmith replied:
Added

LukasPaczos commented on MapboxAndroidDemo/src/global/java/com/mapbox/mapboxandroiddemo/MapboxApplication.java:
```diff
@@ -18,10 +18,13 @@
   public void onCreate() {
     super.onCreate();
     setUpPicasso();
+    FirebaseApp.initializeApp(this);
```
This can also be if-check for debug builds, otherwise, it prints a lot of failure logs in the logcat.

langsmith replied:
Ok, fixed to 

```java
if (!BuildConfig.DEBUG) {
      FirebaseApp.initializeApp(this);
      setUpCrashlytics();
    }
```

langsmith commented on MapboxAndroidDemo/build.gradle:
```diff
@@ -176,4 +183,7 @@ dependencies {
-apply plugin: 'kotlin-android-extensions'
\ No newline at end of file
+apply plugin: 'kotlin-android-extensions'
+if (getGradle().getStartParameter().getTaskRequests().toString().contains("release")) {
```
Fixed

langsmith commented:
> Crashlytics.Builder() instead of the BuildConfig.DEBUG flag check) builds fine for me.

Same for me. This pr is now using `Fabric.with(this, new Crashlytics());`.

> I'm interested in what build issues you were running into with the previous setup? In the current state, using the previous config (Crashlytics.Builder() instead of the BuildConfig.DEBUG flag check) builds fine for me.

Usually a null Crashlytics API key. For example, https://circleci.com/gh/mapbox/mapbox-android-demo/2361. 

![Screen Shot 2019-08-06 at 2 09 22 PM](https://user-images.githubusercontent.com/4394910/62577494-ea0c9400-b853-11e9-96bd-88be273508f5.png)

Removing `ext.enableCrashlytics = false` from `MapboxAndroidDemo/build.gradle` or not having a `Release` check for the `com.google.gms.google-services` plugin, throws the following error/crash:

![Screen Shot 2019-08-06 at 2 13 00 PM](https://user-images.githubusercontent.com/4394910/62577713-5b4c4700-b854-11e9-9617-5c27aa1c7f19.png)


The first three Firebase dependencies use `globalImplementation`
![Screen Shot 2019-08-06 at 3 03 35 PM](https://user-images.githubusercontent.com/4394910/62580593-6191f180-b85b-11e9-9840-2b140c208bcf.png)

However, using `globalImplementation` for

```groovy
dependenciesList.firebaseMessaging
dependenciesList.firebaseRemoteConfig
```

![Screen Shot 2019-08-06 at 2 59 46 PM](https://user-images.githubusercontent.com/4394910/62580362-e03a5f00-b85a-11e9-93b4-a5aa3aafec09.png)

leads to a `> Task :MapboxAndroidDemo:javaPreCompileChinaDebug` crash https://circleci.com/gh/mapbox/mapbox-android-demo/2363

![Screen Shot 2019-08-06 at 3 01 16 PM](https://user-images.githubusercontent.com/4394910/62580474-0d870d00-b85b-11e9-8506-7982e71a7963.png)



Two `productFlavors`, two `buildTypes`, five different Firebase libraries, and debug/release runtime checks, all make for a pretty finegrained setup. Lots of trial & error on local machine and seeing what would pass in CircleCI/Firebase.
